client/db/types.go

```diff
@@ -436,3 +477,126 @@ func RestoreAccountBackup(path string) (*AccountBackup, error) {
+func NewNotification(noteType, subject, details string, severity Severity) Notification {
+	note := Notification{
+		NoteType:    noteType,
+		SubjectText: subject,
+		DetailText:  details,
+		Severeness:  severity,
+	}
+	note.Stamp()
+	return note
+}
+
+// ID is a unique ID based on a hash of the notification data.
+func (n *Notification) ID() dex.Bytes {
+	return hashKey(n.Encode())
+}
+
+// Type is the notification type.
+func (n *Notification) Type() string {
+	return n.NoteType
+}
+
+// Subject is a short description of the notification contents.
+func (n *Notification) Subject() string {
+	return n.SubjectText
+}
+
+// Details should contain more detailed information.
+func (n *Notification) Details() string {
+	return n.DetailText
+}
+
+// Severity is the notification severity.
+func (n *Notification) Severity() Severity {
+	return n.Severeness
+}
+
+// Time is the notification timestamp. The timestamp is set in NewNotification.
+func (n *Notification) Time() uint64 {
+	return n.TimeStamp
+}
+
+// Acked is true if the user has seen the notification. Acknowledgement is
+// recorded with DB.AckNotification.
+func (n *Notification) Acked() bool {
+	return n.Ack
+}
+
+// Stamp sets the notification timestamp. If NewNotification is used to
+// construct the Notification, the timestamp will already be set.
+func (n *Notification) Stamp() {
+	n.TimeStamp = encode.UnixMilliU(time.Now())
+	n.Id = n.ID()
+}
+
+// DBNote is a function to return the *Notification itself. It  should really be
+// defined on the concrete types in core, but is ubiquitous so defined here for
+// convenience.
+func (n *Notification) DBNote() *Notification {
+	return n
+}
+
+// DecodeWallet decodes the versioned blob to a *Wallet.
+func DecodeNotification(b []byte) (*Notification, error) {
+	ver, pushes, err := encode.DecodeBlob(b)
+	if err != nil {
+		return nil, err
+	}
+	switch ver {
+	case 0:
+		return decodeNotification_v0(pushes)
+	}
+	return nil, fmt.Errorf("unknown DecodeNotification version %d", ver)
+}
+
+func decodeNotification_v0(pushes [][]byte) (*Notification, error) {
+	if len(pushes) != 5 {
+		return nil, fmt.Errorf("decodeNotification_v0: expected 5 pushes, got %d", len(pushes))
+	}
+	if len(pushes[3]) != 1 {
+		return nil, fmt.Errorf("decodeNotification_v0: severity push is supposed to be length 1. got %d", len(pushes[2]))
+	}
+
+	return &Notification{
+		NoteType:    string(pushes[0]),
+		SubjectText: string(pushes[1]),
+		DetailText:  string(pushes[2]),
+		Severeness:  Severity(pushes[3][0]),
+		TimeStamp:   intCoder.Uint64(pushes[4]),
+	}, nil
+}
+
+// Encode encodes the Notification to a versioned blob.
+func (n *Notification) Encode() []byte {
+	return dbBytes{0}.
+		AddData([]byte(n.NoteType)).
+		AddData([]byte(n.SubjectText)).
+		AddData([]byte(n.DetailText)).
+		AddData([]byte{byte(n.Severeness)}).
+		AddData(uint64Bytes(n.TimeStamp))
+}
+
+// hashKeySize must be <= 32.
+const hashKeySize = 8
```

Comment:
Why don't we use the full `blake2s.Size`  (32) byes as the key?

Reply:
Primarily size. 32 bytes each adds up, and notifications are not security-critical, so we can tolerate a higher collision rate. It's also a straightforward conversion to `uint64` for places where it makes sense to use an integer for the index instead of a byte array or string.

Reply:
Siphash would maybe be a better overall choice. 8-byte hash size. Decred's GCS filters use siphash. It's also the function used for Python's built-in `hash` function.

Reply:
For a follow-up